src/shell/commands/disk_rebalance.cpp

```diff
@@ -0,0 +1,288 @@
+// Copyright (c) 2017, Xiaomi, Inc.  All rights reserved.
+// This source code is licensed under the Apache License Version 2.0, which
+// can be found in the LICENSE file in the root directory of this source tree.
+
+#include "shell/commands.h"
+#include "shell/argh.h"
+
+#include <math.h>
+#include <fmt/ostream.h>
+#include <dsn/utility/errors.h>
+#include <dsn/utility/output_utils.h>
+#include <dsn/utility/string_conv.h>
+#include <dsn/dist/replication/duplication_common.h>
+
+bool validate_cmd(const argh::parser &cmd,
+                  const std::set<std::string> &params,
+                  const std::set<std::string> &flags)
+{
+    if (cmd.size() > 1) {
+        fmt::print(stderr, "too many params!\n");
+        return false;
+    }
+
+    for (const auto &param : cmd.params()) {
+        if (params.find(param.first) == params.end()) {
+            fmt::print(stderr, "unknown param {} = {}\n", param.first, param.second);
+            return false;
+        }
+    }
+
+    for (const auto &flag : cmd.flags()) {
+        if (params.find(flag) != params.end()) {
+            fmt::print(stderr, "missing value of {}\n", flag);
+            return false;
+        }
+
+        if (flags.find(flag) == flags.end()) {
+            fmt::print(stderr, "unknown flag {}\n", flag);
+            return false;
+        }
+    }
+
+    return true;
+}
+
+std::ostream *get_out_stream(const std::string &file_name)
+{
+    std::ostream *os_ptr = nullptr;
+    if (file_name.empty()) {
+        os_ptr = &std::cout;
+    } else {
+        os_ptr = new std::ofstream(file_name);
+        if (!*os_ptr) {
```

Comment:
What is if (!*os_ptr) mean?
In fds_service.cpp, it use `os_ptr->is_open()` instead.

Reply:
`os_ptr` is `std::ostream`  type (for  `os_ptr = &std::cout`) which has not member function `os_ptr->is_open()`